cmd/storer2elastic/dataprocessor/dataReplayer.go

```diff
@@ -0,0 +1,578 @@
+) (*storer2ElasticData.HeaderData, error) {
+	shardDBInfo, err := getShardDatabaseForEpoch(dbsInfos, epoch, shardInfo.ShardID)
+	if err != nil {
+		return nil, err
+	}
+
+	shardHeader, err := dr.getShardHeader(shardPersisters, shardInfo.HeaderHash)
+	if err != nil {
+		// if new epoch, shard headers can be found in the previous epoch's storage
+		// TODO: for this case return and use all the persisters from the previous epoch in order to find txs
+		shardHeader, err = dr.getFromShardStorer(dbsInfos, shardInfo, epoch-1)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return dr.processHeader(shardPersisters, shardDBInfo, shardHeader, epoch)
+}
+
+func (dr *dataReplayer) getFromShardStorer(dbsInfos []*databasereader.DatabaseInfo, shardInfo *block.ShardData, epoch uint32) (*block.Header, error) {
+	shardDBInfo, err := getShardDatabaseForEpoch(dbsInfos, epoch, shardInfo.ShardID)
+	if err != nil {
+		return nil, err
+	}
+
+	shardPersisters, err := dr.preparePersistersHolder(shardDBInfo)
+	if err != nil {
+		return nil, err
+	}
+
+	defer func() {
+		dr.closePersisters(shardPersisters)
+	}()
+
+	return dr.getShardHeader(shardPersisters, shardInfo.HeaderHash)
+}
+
+func (dr *dataReplayer) getShardHeader(
+	shardPersisters *persistersHolder,
+	hash []byte,
+) (*block.Header, error) {
+	shardHeaderBytes, err := shardPersisters.shardHeadersPersister.Get(hash)
+	if err != nil {
+		return nil, err
+	}
+
+	return dr.headerMarshalizer.UnmarshalShardHeader(shardHeaderBytes)
+}
+
+func (dr *dataReplayer) processHeader(persisters *persistersHolder, dbInfo *databasereader.DatabaseInfo, hdr data.HeaderHandler, epoch uint32) (*storer2ElasticData.HeaderData, error) {
+	miniBlocksHashes := make([]string, 0)
+	shardIDs := dr.getShardIDs()
+	for _, shard := range shardIDs {
+		miniBlocksForShard := hdr.GetMiniBlockHeadersWithDst(shard)
+		for hash := range miniBlocksForShard {
+			miniBlocksHashes = append(miniBlocksHashes, hash)
+		}
+	}
+
+	body, txPool, err := dr.processBodyAndTransactionsPoolForHeader(persisters, miniBlocksHashes)
+	if err != nil {
+		return nil, err
+	}
+
+	return &storer2ElasticData.HeaderData{
+		Header:           hdr,
+		Body:             body,
+		BodyTransactions: txPool,
+	}, nil
+}
+
+func (dr *dataReplayer) getShardIDs() []uint32 {
+	shardIDs := make([]uint32, 0)
+	for shard := uint32(0); shard < dr.shardCoordinator.NumberOfShards(); shard++ {
+		shardIDs = append(shardIDs, shard)
+	}
+	shardIDs = append(shardIDs, core.MetachainShardId)
+
+	return shardIDs
+}
+
+func (dr *dataReplayer) processBodyAndTransactionsPoolForHeader(
+	persisters *persistersHolder,
+	mbHashes []string,
+) (*block.Body, map[string]data.TransactionHandler, error) {
+	txPool := make(map[string]data.TransactionHandler)
+	mbUnit := persisters.miniBlocksPersister
+
+	blockBody := &block.Body{}
+	for _, mbHash := range mbHashes {
+		mbBytes, err := mbUnit.Get([]byte(mbHash))
+		if err != nil {
+			log.Warn("miniblock not found in storage", "hash", []byte(mbHash))
```

Comment:
Unrecoverable error?

Reply:
debatable. in a further refactoring, we can add a bool flag which can tell that we can stop the indexing if something is not found. We'll have to discuss which option is better: stop indexing or index all you can

Reply:
I think there will be an inconsistent state if some txs/miniblocks/blocks are skipped. You would see only part of the transactions to/from your account and the sum of the values in the txs will not be equal to your actual balance. To fix this, a new run will be needed right?

Reply:
The tool should be able to process *everything* . Otherwise should fail and someone will need to resolve the issues. Otherwise, as @raduchis  mentioned we will have an inconsistent indexer database. Maybe without even knowing it.

Reply:
ok, changed the error handling process. if any error occurs, it will be propagated to the top and the app will stop